zero/src/prover.rs

```diff
@@ -131,29 +135,143 @@ impl BlockProverInput {
+                                IndexedStream::from([segment_data]),
+                                &seg_prove_ops,
+                            )
+                            .run(&proof_runtime.heavy_proof)
+                            .await?;
+                            let seg_aggregatable_proof = seg_aggregatable_proof.into_values_sorted().await?;
+                            batch_segment_aggregatable_proofs.extend(seg_aggregatable_proof);
+                        }
+                        segment_counter += 1;
+                    }
+                    debug!(block_number=%block_number, batch=%batch_idx, "finished proving all segments");
+                    // We have proved all the segments in a batch,
+                    // now we need to aggregate them to the batch proof.
+                    // Fold the segment aggregated proof stream into a single batch proof.
+                    let batch_proof = if batch_segment_aggregatable_proofs.len() == 1 {
+                        // If there is only one segment aggregated proof, just transform it to batch proof.
+                        (batch_idx, crate::proof_types::BatchAggregatableProof::from(
+                            batch_segment_aggregatable_proofs.pop().unwrap(),
+                        ))
+                    } else {
+                        Directive::fold(IndexedStream::from(batch_segment_aggregatable_proofs), &seg_agg_ops)
+                            .run(&proof_runtime.light_proof)
+                            .map(move |e| {
+                                e.map(|p| {
+                                    (
+                                        batch_idx,
+                                        crate::proof_types::BatchAggregatableProof::from(p),
+                                    )
+                                })
+                            })
+                            .await?
+                    };
+                    debug!(block_number=%block_number, batch=%batch_idx, "generated batch proof for block");
+                    batch_proof_tx.send(batch_proof).await.context(format!(
+                        "unable to send batch proof, block: {block_number}, batch: {batch_idx}"
+                    ))?;
+                    Ok::<(), anyhow::Error>(())
+                });
+            }
+            // Wait for all the batch proving tasks to finish. Exit early on error.
+            while let Some(it) = batch_proving_futures.next().await {
+                it?;
+            }
+            Ok::<(), anyhow::Error>(())
+        });
+
+        // Collect all the batch proofs.
+        let mut batch_proofs: Vec<(usize, crate::proof_types::BatchAggregatableProof)> = Vec::new();
+        while let Some((batch_idx, batch_proof)) = batch_proof_rx.recv().await {
+            batch_proofs.push((batch_idx, batch_proof));
+        }
+        debug!(block_number=%block_number, "all the batch proofs are collected");
```

Comment:
(to align with the above logging)

```suggestion
        debug!(block_number=%block_number, "collected all batch proofs");
```